### PR TITLE
Drag and drop from export history

### LINKED ISSUE #602: Drag and drop from export history
Ability to drag and drop kaptures directly from the export window

### PR DESCRIPTION
**Pull request by albinekb**
Fixes #602 

I can not get https://electronjs.org/docs/api/native-image#nativeimagecreatefromdataurldataurl to work with the kapture image, any ideas why or how to get the correct image when dragging?

Do we want the screenshot or a static image for all dragged files?

### CODE REVIEW

File: main/export-list.js [hunk 1]
```diff
@@ -193,6 +195,19 @@ class ExportList {
       openEditorWindow(exp.inputPath, exp.originalFps);
     }
   }
+
+  dragExport(createdAt) {
+    track('export/history/drag/recording');
+    const exportsWindow = getExportsWindow();
+
+    const exp = this.exports.find(exp => exp.createdAt === createdAt);
```
Comment (sindresorhus) [inline main/export-list.js]: `exp` => `export`

File: main/export-list.js [hunk 2]
```diff
@@ -2,6 +2,7 @@
 'use strict';
 
 const {dialog} = require('electron');
+const {join: pathJoin} = require('path');
```
Comment (sindresorhus) [inline main/export-list.js]: There's no benefit of using destructuring here. `const path = require('path')` would be better.

File: main/export-list.js [hunk 3]
```diff
@@ -19,6 +20,7 @@ const {toggleExportMenuItem} = require('./menus');
 const Export = require('./export');
 
 const ffmpegPath = util.fixPathForAsarUnpack(ffmpeg.path);
+const fileIconPath = pathJoin(__dirname, '../', 'static/fileIcon.png');
```
Comment (sindresorhus) [inline main/export-list.js]:
```suggestion
const fileIconPath = pathJoin(__dirname, '../static/fileIcon.png');
```

File: main/export-list.js [hunk 4]
```diff
@@ -193,6 +195,19 @@ class ExportList {
...
   }
+
+  dragExport(createdAt) {
+    track('export/history/drag/recording');
+    const exportsWindow = getExportsWindow();
+
+    const exp = this.exports.find(exp => exp.createdAt === createdAt);
+    if (exp) {
+      exportsWindow.webContents.startDrag({
+        file: exp.context.targetFilePath,
```
[the remark this answers is not in the document]
Comment (sindresorhus) [reply main/export-list.js]: Why do you think it's incorrect? (I haven't really looked into this yet)

**Comment by sindresorhus** [comment]
> Do we want the screenshot or a static image for all dragged files?

Screenshot

File: main/export-list.js (hunk 1 above)
Comment (albinekb) [reply main/export-list.js]: This is copy-pasted from above, should i change that too? or make a function to get an export by createdAt?

File: main/export-list.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (albinekb) [reply main/export-list.js]: It seems to only be set by the default file export share plugin, see:
https://github.com/wulkano/kap/blob/d021f0cc1e936782928c12501ea4c733e9aaa4ee/main/export-list.js#L110-L131

File: main/export-list.js (hunk 2 above)
Comment (albinekb) [reply main/export-list.js]: I agree 👍

File: main/export-list.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (karaggeorge) [reply main/export-list.js]: Yeah, the rest of the plugins only use the temp file(s). Only the default plugin actually copies it to the "target file path"

[the remark this answers is not in the document]
Comment (karaggeorge) [reply main/export-list.js]: Pretty sure you can use `export.filePath` to get the path of the file after the conversion. It gets the value after the conversion is completed so we would have to make sure you can't drag/drop before it's done:

https://github.com/wulkano/kap/blob/c0baee3659b97e6e9c38868416ab4748cb1a3646/main/export.js#L111

[the remark this answers is not in the document]
Comment (albinekb) [reply main/export-list.js]: Let's try that 👍 https://superuser.com/questions/187071/in-macos-how-often-is-tmp-deleted

>Short answer: by default, files that aren't accessed in three days are deleted from /tmp

File: main/export-list.js (hunk 1 above)
Comment (sindresorhus) [reply main/export-list.js]: > or make a function to get an export by createdAt?

👍

Comment (albinekb) [reply main/export-list.js]: `export`  seems to be restricted keyword?
image
image

Comment (karaggeorge) [reply main/export-list.js]: I remember that haha. Can't use export as a variable name because it's reserved

**Comment by albinekb** [comment]
@sindresorhus I fixed your comments. We can't use `export` as the variable name as it's restricted. 

I can't get the screenshot to work, any ideas on where i should get the sceenshot image to show when dragging?

The only thing i got to work was generating the screenshot just when you start dragging, which causes the drag to appear stuttering in the beginning. When I am using the same screenshot data uri with electron's api to convert it to an NativeImage, the app crashes when dragging the screenshot.

File: main/export-list.js (hunk 1 above)
Comment (sindresorhus) [reply main/export-list.js]: Ugh, I hate that. In Swift, you can at least wrap the variable name in backticks to allow using reserved names.

Comment (sindresorhus) [reply main/export-list.js]: `exp` => `exportItem` then?

Comment (albinekb) [reply main/export-list.js]: yep, that's better

**Comment by sindresorhus** [comment]
Closing in favor of #641.

**Pull request closed without merging**